samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -0,0 +1,172 @@
+   * <p>
+   * <b>Note:</b>This identifier has to be unique across the instances of StreamProcessors.
+   */
+  private static final String PROCESSOR_ID = "processor.id";
+  private final int processorId;
+  private final JobCoordinator jobCoordinator;
+  private final SamzaContainerController containerController;
+
+  /**
+   * Create an instance of StreamProcessor that encapsulates a JobCoordinator and Samza Container
+   * <p>
+   * JobCoordinator controls how the various StreamProcessor instances belonging to a job coordinate. It is also
+   * responsible generating and updating JobModel.
+   * When StreamProcessor starts, it starts the JobCoordinator and brings up a SamzaContainer based on the JobModel.
+   * SamzaContainer is executed using an ExecutorService.
+   * <p>
+   * <b>Note:</b> Lifecycle of the ExecutorService is fully managed by the StreamProcessor, and NOT exposed to the user
+   *
+   * @param processorId            Unique identifier for a processor within the job. It has the same semantics as
+   *                               "containerId" in Samza
+   * @param config                 Instance of config object - contains all configuration required for processing
+   * @param customMetricsReporters Map of custom MetricReporter instances that are to be injected in the Samza job
+   * @param asyncStreamTaskFactory The {@link AsyncStreamTaskFactory} to be used for creating task instances.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                         AsyncStreamTaskFactory asyncStreamTaskFactory) {
+    this(processorId, config, customMetricsReporters, (Object) asyncStreamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the provided {@link StreamTaskFactory}.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                         StreamTaskFactory streamTaskFactory) {
+    this(processorId, config, customMetricsReporters, (Object) streamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the "task.class" configuration instead of a task factory.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters) {
+    this(processorId, config, customMetricsReporters, (Object) null);
+  }
+
+  private StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                          Object taskFactory) {
+    this.processorId = processorId;
+
+    Map<String, String> updatedConfigMap = new HashMap<>();
+    updatedConfigMap.putAll(config);
+    updatedConfigMap.put(PROCESSOR_ID, String.valueOf(this.processorId));
```

Comment:
Yeah. It is a little tricky to understand why this is needed. "processorId" is equivalent to "containerId" . However, containerId (in today's Yarn world) can only be between 0 to N-1 , where N is defined by job.container.count. This is not the case for processorId. It is directly defined through the StreamprocessorAPI by the user application. So, the grouper - SspGrouper and TaskNameGrouper needs to know the processorId in order to use it as a  key in the JobModel definition.
If you take a look at AllSspToSingleTaskGrouperFactory and SingleContainerGrouperFactory, you will get an idea of how this processorId is used in these groupers.
Imo, the correct long-term solution should be to change the grouper interface to take in specific parameters, instead of an arbitrary bag of key/values pairs. Since it is kind of orthogonal to the scope of this work, the change to the interface has not yet been made.